src/forklift/slack.py

```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env python
+# * coding: utf8 *
+'''
+slack.py
+A module that holds the constructs for using the slack api
+'''
+
+import math
+from abc import ABC, abstractmethod
+from datetime import datetime
+from enum import Enum
+from json import dumps
+
+
+MAX_BLOCKS = 50
+MAX_CONTEXT_ELEMENTS = 10
+MAX_SECTION_FIELD_ELEMENTS = 10
+MAX_LENGTH_SECTION = 3000
+MAX_LENGTH_SECTION_FIELD = 2000
+
+def split(arr, size):
+    result = []
+    while len(arr) > size:
+        piece = arr[:size]
+        result.append(piece)
+        arr = arr[size:]
+
+    result.append(arr)
+
+    return result
+
+
+def lift_report_to_blocks(report):
+    '''turns the forklift lift report object into blocks
+    '''
+    message = Message()
+
+    message.add(SectionBlock(f':tractor:       :package: *Forklift Lift Report* :package:      :tractor:'))
+
+    percent = report['num_success_pallets'] / report['total_pallets'] * 100
+    if percent == 100:
+        percent = ':100:'
+    else:
+        percent = f'{str(math.floor(percent))}% success'
+
+    message.add(
+        ContextBlock([
+            f'*{datetime.now().strftime("%B %d, %Y")}*',
+            report['hostname'],
+            f'*{report["num_success_pallets"]}* of *{report["total_pallets"]}* pallets ran successfully',
+            f'{percent}',
+            f'total time: *{report["total_time"]}*',
+        ])
+    )
+
+    message.add(DividerBlock())
+
+    if report['git_errors']:
+        git_block = SectionBlock('git errors')
+
+        for error in report['git_errors']:
+            git_block.fields.append(error)
+
+        message.add(git_block)
+
+    if report['import_errors']:
+        import_block = SectionBlock('python import errors')
+
+        for error in report['import_errors']:
+            import_block.fields.append(error)
+
+        message.add(import_block)
+
+    for pallet in report['pallets']:
+        success = ':fire:'
+
+        if pallet['success']:
+            success = ':heavy_check_mark:'
+
+        message.add(SectionBlock(f'{success} *{pallet["name"].split(":")[2]}*'))
+        message.add(ContextBlock([f'{pallet["total_processing_time"]}{"  |  " + pallet["message"] if pallet["message"] else ""}']))
+
+        crate_elements = []
+
+        for crate in pallet['crates']:
+            show_message = False
+            if crate['result'] in ['Data updated successfully.', 'Created table successfully.', 'No changes found.']:
```

Comment:
Could these strings be imported from models.py?

Reply:
Yup